### PR TITLE
feat: support `LargeList` in `cardinality`

### PR DESCRIPTION
**Pull request by Weijun-H**
## Which issue does this PR close?

Parts #8185

### CODE REVIEW

File: datafusion/physical-expr/src/array_expressions.rs [hunk 1]
```diff
@@ -2110,16 +2110,38 @@ pub fn cardinality(args: &[ArrayRef]) -> Result<ArrayRef> {
...
-    let result = list_array
-        .iter()
-        .map(|arr| match compute_array_dims(arr)? {
-            Some(vector) => Ok(Some(vector.iter().map(|x| x.unwrap()).product::<u64>())),
-            None => Ok(None),
-        })
-        .collect::<Result<UInt64Array>>()?;
-
+    let result = match &args[0].data_type() {
+        DataType::List(_) => {
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/array_expressions.rs:2114]: Although this implementation is small, I think having a closure for generic array is still a good practice.

Comment (alamb) [reply datafusion/physical-expr/src/array_expressions.rs:2114]: I agree it would be better to avoid the duplication if possible

**Review by jayzhan211 — APPROVED** [review]
LGTM

File: datafusion/physical-expr/src/array_expressions.rs [hunk 2]
```diff
@@ -2110,16 +2110,34 @@ pub fn cardinality(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+    match &args[0].data_type() {
+        DataType::List(_) => {
+            let list_array = as_list_array(&args[0])?;
+            generic_list_cardinality::<i32>(list_array)
+        }
+        DataType::LargeList(_) => {
+            let list_array = as_large_list_array(&args[0])?;
+            generic_list_cardinality::<i64>(list_array)
+        }
+        _ => {
```
Comment (comphead) [inline datafusion/physical-expr/src/array_expressions.rs:2122]:
```suggestion
                other => {
            exec_err!(
                "cardinality does not support type '{:?}'",
                other
            )
        }
```

**Review by alamb — APPROVED** [review]
Looks good to me -- thanks @Weijun-H 

I think @comphead  had a suggestion for improvement, but we could also do it as a follow on PR as well

**Review by comphead — APPROVED** [review]
lgtm thanks @Weijun-H

**Pull request merged**